Sprint9/project

RyanZomparelli opened the pull request:
What I Did:
This pull request introduces major enhancements to the app's server integration.

Key changes include:

1. Add Api class to manage all server requests (GET, POST, PATCH, DELETE) in a centralized, reusable way.

2. Implement avatar editing functionality including:

   --New edit avatar popup and form

   --Button UI and logic per Figma spec

3. Add new delete card modal with complete logic and styles

4. Wire up server to:

   --Load initial cards

   --Add new cards

   --Edit profile info

   --Delete cards

   --Like/unlike cards

   --Update avatar

RyanZomparelli committed:
feat: Add Api class for server data fetching and rendering

RyanZomparelli committed:
fix(index.js) import resetValidation to resolve ReferenceError on line 150

RyanZomparelli committed:
chore(Api.js): remove redundant authorization property from the class constructor

RyanZomparelli committed:
refactor(index.js) use array destructuring for API response data

RyanZomparelli committed:
refactor: update editProfileSubmit handler to save user input to server and update DOM

- Add new Api method `editUserInfo` to send PATCH request with updated profile data.
- Update index.js to call this method and update DOM with returned user info.
- Import `disableSubmitButton` from validation.js to resolve ReferenceError on form submit.

RyanZomparelli committed:
refactor: use object destructuring in editUserInfo and handler for improved clarity

RyanZomparelli committed:
refactor: upodate hadleAddCardSubmit to save new cards to the server and update DOM

-Add new API method 'addNewCard' to send POST request with new card data.
-Update index.js to call this method and update the DOM with returned card data.

RyanZomparelli committed:
refactor(Api.js): implement response handler in Api.js to conform to DRY priciples

RyanZomparelli committed:
chore(index.html): remove hardcoded profile values

RyanZomparelli committed:
refactor: create a new delete card confirmation modal with styles and logic.

Add a new modal and form to index.js for deleting cards and, implement the modal styles from figma.
Add two new functions to handle card deletion in index.js.
Add a method in Api.js to send DELETE requests.

RyanZomparelli committed:
refactor: add two new Api methods for saving card likes and implement them in index.js

RyanZomparelli committed:
feat: add an edit avatar button with styles from the figma design and adjust the mobile layout for the delete card modal

RyanZomparelli committed:
feat: add new edit avatar popup and form and add click listener to edit avatar button

RyanZomparelli committed:
feat:
- Wire up open and close logic to edit avatar button and add new Api method to update avatar data.
- Improve UI with submitBtn function.

RyanZomparelli committed:
fix(index.js): fix card ordering issue by appendng the data upon page load

RyanZomparelli committed:
chore: move Api.js to a new utils directory and organize index.js

RyanZomparelli committed:
chore: update README.md

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -198,17 +194,164 @@ function getCardElement(data) {
+  api
+    .editAvatar(avatarInput.value)
+    .then((data) => {
+      handleSubmitBtnUI(editAvatarSubmitBtn);
```
According to the project requirements, the submit button text should update **as soon as the form is submitted**, while waiting for the server response.

Currently, it's being updated after a successful response, which is not ideal. To improve the user experience and align with the spec, the text should change before the API call and reset to its original state after the request completes, whether successful or not. This should be handled in a `finally` block.

Please refactor the handler function like this:

```js
function handleNewAvatarSubmit() {
  handleSubmitBtnUI(editAvatarSubmitBtn, true); // Show loading state
  api
    .editAvatar(avatarInput.value)
    .then((data) => {
      profileAvatar.src = data.avatar;
      closeModal(editAvatarModal);
      newAvatarForm.reset();
      disableSubmitButton(editAvatarSubmitBtn, settings);
    })
    .catch((error) => console.error(error))
    .finally(() => handleSubmitBtnUI(editAvatarSubmitBtn, false)); // Reset UI
}
```

Note: You’ll also need to update the `handleSubmitBtnUI` function, please see my comment there for the required changes.

Lastly, make sure to apply this pattern consistently across all form submission handlers in your project, including the delete confirmation form.

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -198,17 +194,164 @@ function getCardElement(data) {
+      handleSubmitBtnUI(editAvatarSubmitBtn);
+      profileAvatar.src = data.avatar;
+      closeModal(editAvatarModal);
+      disableSubmitButton(editAvatarSubmitBtn, settings);
```
You need to reset the form after a successful response from the server.

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -103,61 +104,35 @@ function handleKeyDown(e) {
 
-editModalCloseBtn.addEventListener("click", () => closeModal(editProfileModal));
-previewCloseBtn.addEventListener("click", () => closeModal(previewModal));
+function handleSubmitBtnUI(btn) {
```
Instead of relying on the button's `disabled` property to determine when to update its text, it's better to pass an explicit `loading` boolean as a parameter. This makes the logic clearer and more predictable:

```js
function handleSubmitBtnUI(btn, loading) {
  if (loading) {
    btn.textContent = "Saving...";
  } else {
    btn.textContent = "Save";
  }
}
```

This approach gives you more control over the button's state and helps separate concerns. The `loading` state is now handled explicitly, rather than being inferred from unrelated properties like `disabled`.

sebasLTripleTen reviewed (changes requested):
Hello, Ryan!

Thank you for submitting your project. It's clear that a lot of work has been put into this project.

**Parts I really liked in your project:**

- Well-structured and clean code
- All variables, functions, and handlers are in the right places
- The code for closing the modal windows is executed in the right place
- You've created the class instances and used its methods correctly
- Great job handling potential errors at the end of your server requests.

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

- All comments in the code, along with the ones below:
    - You should reset the Edit Avatar form after a successful response from the server.
    - The submit button text should update as soon as the form is submitted, while waiting for the server response. Currently, it's being updated after a successful response.

There are also some comments for improvement marked as **"Could be improved"**. They don’t mean that you’ve done anything wrong, but they’re things we wanted to point out that can help you further develop your skills.

**Reminder:** For the project to be accepted, you must fix all **"Needs correcting"** comments. **"Could be improved"** comments are optional, so you can implement them if you choose.

It's fantastic to see how far you've come with this project. Even though there are areas for improvement, your hard work and dedication are evident. Remember that each step forward is progress. If you need help with any part of the feedback, our tutors are available to support you. Keep pushing forward and continue your impressive journey!

RyanZomparelli committed:
fix: refactor handleSubmitBtnUI function logic and timing and update README.md

- Refactor handleSubmitBtnUI to utilize a isLoading argument to be able to accomodate the delete modal.
- Call the handleSubmitBtnUI once before the api call and then again in a .finally method call to reset the button state even if the api call fails.
- Wrap the UI function in a setTimeout function to prevent rapid flashing of UI states.
- Update the OOP note in the README.md

RyanZomparelli committed:
refactor(index.js): reset the form fields after adding a new avatar and adjust disableSubmitBtn timing

RyanZomparelli committed:
chore(README.md): Update video presentation

sebasLTripleTen reviewed (approved):
You’ve done a great job!

Your project has been accepted.

Good luck!
